File: ProblemSolve/1305.cpp
/*******************************************
 * All Elements in Two Binary Search Trees *
 *******************************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void getElementsAscending(TreeNode* root, vector<int>& list){
        if(root == NULL){
            return;
        }
        
        if(root -> left != NULL){
            getElementsAscending(root -> left, list);
        }
        
        list.push_back(root -> val);
        
        if(root -> right != NULL){
            getElementsAscending(root -> right, list);
        }
        
        return;
    }
    
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        vector<int> listForRoot1;
        vector<int> listForRoot2;
        vector<int> ans;
        
        getElementsAscending(root1, listForRoot1);
        getElementsAscending(root2, listForRoot2);
        
        int len1 = listForRoot1.size();
        int len2 = listForRoot2.size();
        int idx1 = 0;
        int idx2 = 0;
        
        while(idx1 < len1 && idx2 < len2){
            if(listForRoot1[idx1] >= listForRoot2[idx2]){
                ans.push_back(listForRoot2[idx2]);
                idx2++;
            }
            else{
                ans.push_back(listForRoot1[idx1]);
                idx1++;
            }
        }
     
        while(idx1 < len1){
            ans.push_back(listForRoot1[idx1]);
            idx1++;
        }
        
        while(idx2 < len2){
            ans.push_back(listForRoot2[idx2]);
            idx2++;
        }
        
        return ans;
    }
};

File: ProblemSolve/1305.cs
/*******************************************
 * All Elements in Two Binary Search Trees *
 *******************************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public void GetElementsList(TreeNode root, ref List<int> list){
        if(root == null){
            return;
        }
        
        if(root.left != null){
            GetElementsList(root.left, ref list);
        }
        
        list.Add(root.val);
        
        if(root.right != null){
            GetElementsList(root.right, ref list);
        }
        
        return;
    }
    
    public IList<int> GetAllElements(TreeNode root1, TreeNode root2) {
        List<int> root1List = new List<int>();
        List<int> root2List = new List<int>();
        List<int> ans = new List<int>();
        
        GetElementsList(root1, ref root1List);
        GetElementsList(root2, ref root2List);
        
        int idx1 = 0, idx2 = 0, len1 = root1List.Count, len2 = root2List.Count;
        
        while(idx1 < len1 && idx2 < len2){
            if(root1List[idx1] >= root2List[idx2]){
                ans.Add(root2List[idx2]);
                idx2++;
            }
            else{
                ans.Add(root1List[idx1]);
                idx1++;
            }
        }
        
        while(idx1 < len1){
            ans.Add(root1List[idx1++]);            
        }
        
        while(idx2 < len2){
            ans.Add(root2List[idx2++]);
        }
        
        return ans;        
    }
}